packages/contracts-hyperlane/src/token/getTokenRouterDeployTransactions.ts

```diff
@@ -0,0 +1,156 @@
+import { HypNative } from "../artifacts/HypNative.js";
+import { TransparentUpgradeableProxy } from "../artifacts/TransparentUpgradeableProxy.js";
+import { HypERC20 } from "../artifacts/HypERC20.js";
+
+const contractExists = async (client: Client<Transport, Chain, Account>, address: Address) => {
+    const code = await getAction(client, getCode, "getCode")({ address });
+    return !!code;
+};
+
+export async function getTokenRouterDeployTransactions(
+    client: Client<Transport, Chain, Account>,
+    parameters: {
+        owner?: Address;
+        mailboxAddress: Address;
+        collateralAddress?: Address;
+        tokenType: TokenTypeExtended;
+        tokenMetadata?: { name: string; symbol: string; decimals: number };
+        totalSupply?: bigint;
+    },
+) {
+    const owner = parameters?.owner ?? client.account.address;
+    const { tokenType, mailboxAddress, tokenMetadata } = parameters;
+
+    const transactions: { to: Address; data: Hex }[] = [];
+
+    let tokenRouterImplAddress: Address;
+    let tokenRouterImplDeployData: { salt: Hash; bytecode: Hex };
+
+    let tokenRouterProxyAddress: Address;
+    let tokenRouterProxyDeployData: { salt: Hash; bytecode: Hex };
+
+    const proxyAdminDeployArgs = {
+        salt: zeroHash,
+        bytecode: ProxyAdmin.bytecode,
+        initData: encodeFunctionData({ abi: Ownable.abi, functionName: "transferOwnership", args: [owner] }),
+    };
+    const proxyAdminAddress = getDeployAddress(client.account.address, proxyAdminDeployArgs);
+
+    const proxyAdminExists = await contractExists(client, proxyAdminAddress);
+
+    if (!proxyAdminExists) {
+        transactions.push(getDeployFunctionData(owner, [proxyAdminDeployArgs]));
+    }
+
+    switch (tokenType) {
+        case TokenTypeExtended.native:
+            tokenRouterImplDeployData = {
+                salt: zeroHash,
+                bytecode: encodeDeployData({
+                    abi: HypNative.abi,
+                    bytecode: HypNative.bytecode,
+                    args: [mailboxAddress],
+                }),
+            };
+            tokenRouterImplAddress = getDeployDeterministicAddress(tokenRouterImplDeployData);
+
+            tokenRouterProxyDeployData = {
+                salt: zeroHash,
+                bytecode: encodeDeployData({
+                    abi: TransparentUpgradeableProxy.abi,
+                    bytecode: TransparentUpgradeableProxy.bytecode,
+                    args: [
+                        tokenRouterImplAddress,
+                        proxyAdminAddress,
+                        encodeFunctionData({
+                            abi: HypNative.abi,
+                            functionName: "initialize",
+                            args: [zeroAddress, zeroAddress, owner],
+                        }),
+                    ],
+                }),
+            };
+
+            tokenRouterProxyAddress = getDeployDeterministicAddress(tokenRouterProxyDeployData);
```

Comment:
Idk if the proxy should be deployed this way.
I think better to use Create2Factory and enforce msgSender to avoid any risk of someone deploying to same address without owner's consent.

Reply:
Why?

Reply:
Writing summary of discussion.

"Counterfactual address ownership" refers to who can deploy (using CREATE2) the code that lives on a counterfactual address.

When deploying with arachnid deterministic deployer, anyone can replay the same deploy call on any chain. As such anyone can redeploy the same bridge token (with the original owner) on any chain. This is not a severe vulnerability per-se (owner is still same) but might not be a desirable feature (owner may want to determine when official bridge token address is live on a chain).

Create2Factory solves this by introducing optional msgSender enforcement into the CREATE2 salt. This means only the original deployer can redeploy to the same address.